README
Pootle: a web translation and translation management engine
===========================================================

* How it works
* References
* Requirement/Installation
* Running
* Bug reporting
* Notes
* Assignments

How it works
------------
The concept is: a PO file is a document, so lets use PO files directly, 
rather than storing it in a database.  But we need quick access to statistics 
so we cache them in .stats files. This lets us do most checks etc in advance 
and then iterate quickly.

References
----------
Bugzilla: http://bugs.wordforge.org/
Wiki: http://translate.sourceforge.net/wiki/pootle
Mailing List: https://lists.sourceforge.net/lists/listinfo/translate-pootle
	translate-pootle@lists.sourceforge.net
Mail Archives: http://sourceforge.net/mailarchive/forum.php?forum=translate-pootle

Requirements/Installation
-------------------------
Please read http://translate.sourceforge.net/wiki/pootleadmin for the lastest instructions.

Pootle requires the Translate Toolkit package from 
http://translate.sourceforge.net as it uses most of the other modules.

The Translate Toolkit requires Python 2.2 or higher, but Pootle itself 
probably only works with Python 2.3 or higher.  The Translate Toolkit must 
be installed properly i.e. it needs to be accessible in your Python Path. This
will normally be accomplished by installing from source:

	python setup.py install

For the web interface, Pootle requires jToolkit (http://jtoolkit.sourceforge.net/)
jToolkit can run applications either via Apache using mod_python, or standalone 
from the command line.  Pootle works best at the moment being run from the 
command line since we don't handle inter-process locking etc.  It's also 
much easier! To install jToolkit run:

	tar xvzf jToolkit-N.N.N.tar.gz
	cd jToolkit-N.N.N
	python jToolkitSetup.py install

From version 0.9 of Pootle, kid (http://kid.lesscode.org/), and elementtree
(http://effbot.org/downloads/#elementtree) are also needed.

Installation of PyLucene is optional, but helps to speed up searching.

To install Pootle itself, run

	tar xzf Pootle-N.N.N.tar.gz
	cd Pootle-N.N.N
	python pootlesetup.py install

Running
-------
To run, edit pootle.prefs and specify your languages and projects. The key detail 
here is the podir for each project.  Then from the Pootle directory run:

	PootleServer

Use --help to see the other options.  You can tell it which prefs file to use, 
what HTML file and what port to run on (the default is 8080). The defaults will 
generally work.

Now visit http://localhost:8080/ and begin translating.

To stop, press Ctrl-C. On Windows, it will only react to the Ctrl-C after a webpage is requested.

Bug Reporting/Feature requests
------------------------------
You can always report bugs or feature requests on the mailing list but because of the increase
in users and the fact that bug reports do go missing it is probably best to place your
bug report in Bugzilla: http://bugs.wordforge.org/

If you have a traceback or a patch then please include it.  Also please be quite specific
about how the bug occured and what you expected to happen.

If this is a feature request then try to be specific about how you think this feature should 
work.

Notes
-----
Files should be reindexed automatically.
To reindex manually, simply remove any stats files you want reindexed, or run 
	python pootle.py --refreshstats

Assignments
-----------
To assign sstrings to users, you need to create a xxx.po.assigns file in the same 
directory as xxx.po This should contain lines in the following format:

username: action: 0,2-34,36,99

where action can be 'suggest' or 'review', and the numbers refer to 0-based item 
numbers in the file.



ChangeLog
2006-06-20 08:45  dwaynebailey

	* README: [Forwardport] update dependencies

2006-06-19 09:06  friedelwolff

	* translatepage.py: Fix traceback when searching for nonexisting
	  text

2006-06-19 08:59  friedelwolff

	* templates/pootlepage.html: Add link tags (stylesheet, favicon),
	  Add Content-Type

2006-06-15 16:32  dwaynebailey

	* PootleServer, __init__.py, adminpages.py, benchmark.py,
	  conflict2suggest.py, filelocations.py, indexpage.py,
	  pagelayout.py, pootle.py, pootlefile.py, potree.py, projects.py,
	  translatepage.py, users.py, versioncontrol.py: [Forwardport]
	  Update copyright dates and add missing copyright headers to
	  files.

2006-06-15 16:11  dwaynebailey

	* pootle.py: Makre sure that you have the correct version of
	  Translate Toolkit before you can run PootleServer.  Bump the
	  toolkit build number up.

2006-06-15 09:52  dwaynebailey

	* createpootlepot: Make it clear the Zuza is Translate.org.za

2006-06-15 01:18  dwaynebailey

	* templates/projectadmin.html: [Forwardport] Remove proprietary
	  attribute from select tag

2006-06-15 00:55  dwaynebailey

	* templates/: fileindex.html, language.html, project.html: <script>
	  tags must have a type= attribute

2006-06-14 18:05  friedelwolff

	* adminpages.py, templates/adminusers.html: Fix editing of user
	  admin interface

2006-06-14 15:52  friedelwolff

	* pootle.py: Enable caching for CSS, JS, and images

2006-06-14 00:44  friedelwolff

	* translatepage.py: Another capitalisation fix skip->Skip

2006-06-13 14:49  friedelwolff

	* indexpage.py: Temporary fix for link at multigoal file

2006-06-13 14:21  friedelwolff

	* indexpage.py: Reword "No Goal" -> "Not in a goal"

2006-06-13 13:39  dwaynebailey

	* templates/: fileindex.html, options.html, projectlangadmin.html:
	  [FOrwardport] Remove type= attributes from tags (img, td, etc)

2006-06-13 13:27  dwaynebailey

	* templates/: adminindex.html, adminlanguages.html,
	  adminprojects.html, adminusers.html, language.html, navbar.html,
	  project.html, translatepage.html: [Forwardport] Remove type=
	  attributes from tags (img, td, etc)

2006-06-13 13:06  dwaynebailey

	* templates/login.html: [Forwardport] Remove type= attributes from
	  tags (img, td, etc)

2006-06-13 13:00  dwaynebailey

	* templates/activate.html: [Forwardport] Remove type= attributes
	  from tags (img, td, etc)

2006-06-13 12:56  dwaynebailey

	* templates/register.html: [Forwardport] Remove type= attributes
	  from tags (img, td, etc)

2006-06-13 12:27  dwaynebailey

	* templates/pootlepage.html: [Forwardport] Remove type= attributes
	  from tags (img, td, etc)

2006-06-13 12:15  friedelwolff

	* translatepage.py: Make capitalisation consitent

2006-06-13 11:07  dwaynebailey

	* pagelayout.py, templates/navbar.html: [Forwardport] Change cls ->
	  class

2006-06-13 10:55  dwaynebailey

	* templates/pootlepage.html: [Forwardport] Height is not a valid
	  attribute, change it to a style attribute.

2006-06-13 10:41  dwaynebailey

	* templates/pootlepage.html: [Forwardport] Only show 'logged in
	  as...' if a session is actually open.

2006-06-13 10:14  dwaynebailey

	* indexpage.py, templates/fileindex.html, templates/navbar.html,
	  templates/translatepage.html: [Forwardport] Make all empty tags
	  ie <br>, <img>, etc end in [space]/> remove lonely </img> if not
	  needed.

2006-06-13 08:43  friedelwolff

	* projects.py: Support for non-ascii goalnames

2006-06-12 10:54  friedelwolff

	* potree.py, projects.py: Detect gnu style projects differently

2006-06-09 21:59  friedelwolff

	* test_client.py: Fix unit url passed and expected

2006-06-09 21:57  friedelwolff

	* test_client.py: Adapt tests to new pofile output without the
	  unnecessary newline at the end

2006-06-09 20:02  friedelwolff

	* test_client.py: Fix unit test (a PO header ofsets the part that
	  is tested for)

2006-06-08 14:26  friedelwolff

	* templates/translatepage.html: Don't use HTML comment for
	  JavaScript, lest kid ignores it

2006-06-08 14:16  friedelwolff

	* templates/translatepage.html: Make resizing javascript work again

2006-06-08 12:42  friedelwolff

	* translatepage.py: Decode special characters if necessary

2006-06-08 12:21  friedelwolff

	* templates/translatepage.html: [Forwardport] Show special
	  characters again, make copy button work again

2006-06-05 07:16  friedelwolff

	* pagelayout.py: Internationalise string-statistics

2006-06-05 07:07  friedelwolff

	* templates/register.html: Internationalise one last string

2006-06-02 16:03  friedelwolff

	* createpootlepot: Specify outputfile

2006-06-02 14:18  friedelwolff

	* adminpages.py, templates/adminlanguages.html,
	  templates/adminprojects.html, templates/adminusers.html,
	  templates/projectlangadmin.html: Internationalise admin pages

2006-06-02 13:21  friedelwolff

	* adminpages.py, templates/adminindex.html: Internationalise main
	  admin page

2006-06-02 11:32  friedelwolff

	* pagelayout.py, templates/pootlepage.html: Internationalise side
	  links

2006-06-02 11:20  friedelwolff

	* templates/languages.html: Use localised aboutlink

2006-06-02 10:08  friedelwolff

	* createpootlepot: Create pot file with "l10n" as automatic comment
	  keyword

2006-06-02 09:40  friedelwolff

	* templates/pootlepage.html, pagelayout.py: Resolve banner vs. logo
	  confusion

2006-06-02 09:18  friedelwolff

	* potree.py: Encode special characters before comparison

2006-06-01 20:11  friedelwolff

	* potree.py: Format new language code correctly as la_CO, allow

pootle.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# 
# Copyright 2004-2006 Zuza Software Foundation
# 
# This file is part of translate.
#
# translate is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
# 
# translate is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with translate; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

from jToolkit.web import server
from jToolkit.web import templateserver
from jToolkit.web import session
from jToolkit import prefs
from jToolkit import localize
from jToolkit.widgets import widgets
from jToolkit.widgets import spellui
from jToolkit.widgets import thumbgallery
from jToolkit.web import simplewebserver
from Pootle import indexpage
from Pootle import adminpages
from Pootle import translatepage
from Pootle import pagelayout
from Pootle import projects
from Pootle import potree
from Pootle import users
from Pootle import filelocations
from Pootle import __version__ as pootleversion
from translate import __version__ as toolkitversion
from jToolkit import __version__ as jtoolkitversion
import sys
import os
import random
import pprint

class PootleServer(users.OptionalLoginAppServer, templateserver.TemplateServer):
  """the Server that serves the Pootle Pages"""
  def __init__(self, instance, webserver, sessioncache=None, errorhandler=None, loginpageclass=users.LoginPage):
    if sessioncache is None:
      sessioncache = session.SessionCache(sessionclass=users.PootleSession)
    self.potree = potree.POTree(instance)
    super(PootleServer, self).__init__(instance, webserver, sessioncache, errorhandler, loginpageclass)
    self.templatedir = filelocations.templatedir
    self.setdefaultoptions()

  def loadurl(self, filename, context):
    """loads a url internally for overlay code"""
    # print "call to load %s with context:\n%s" % (filename, pprint.pformat(context))
    filename = os.path.join(self.templatedir, filename+os.extsep+"html")
    if os.path.exists(filename):
      return open(filename, "r").read()
    return None

  def saveprefs(self):
    """saves any changes made to the preferences"""
    # TODO: this is a hack, fix it up nicely :-)
    prefsfile = self.instance.__root__.__dict__["_setvalue"].im_self
    prefsfile.savefile()

  def setdefaultoptions(self):
    """sets the default options in the preferences"""
    changed = False
    if not hasattr(self.instance, "title"):
      setattr(self.instance, "title", "Pootle Demo")
      changed = True
    if not hasattr(self.instance, "description"):
      defaultdescription = "This is a demo installation of pootle. The administrator can customize the description in the preferences."
      setattr(self.instance, "description", defaultdescription)
      changed = True
    if not hasattr(self.instance, "baseurl"):
      setattr(self.instance, "baseurl", "/")
      changed = True
    if changed:
      self.saveprefs()

  def changeoptions(self, argdict):
    """changes options on the instance"""
    for key, value in argdict.iteritems():
      if not key.startswith("option-"):
        continue
      optionname = key.replace("option-", "", 1)
      setattr(self.instance, optionname, value)
    self.saveprefs()

  def inittranslation(self, localedir=None, localedomains=None, defaultlanguage=None):
    """initializes live translations using the Pootle PO files"""
    self.localedomains = ['jToolkit', 'pootle']
    self.localedir = None
    self.languagelist = self.potree.getlanguagecodes('pootle')
    self.languagenames = self.potree.getlanguages()
    self.defaultlanguage = defaultlanguage
    if self.defaultlanguage is None:
      self.defaultlanguage = getattr(self.instance, "defaultlanguage", "en")
    if self.potree.hasproject(self.defaultlanguage, 'pootle'):
      try:
        self.translation = self.potree.getproject(self.defaultlanguage, 'pootle')
        return
      except:
        self.errorhandler.logerror("Could not initialize translation")
    # if no translation available, set up a blank translation
    super(PootleServer, self).inittranslation()

  def gettranslation(self, language):
    """returns a translation object for the given language (or default if language is None)"""
    if language is None:
      return self.translation
    else:
      try:
        return self.potree.getproject(language, 'pootle')
      except:
        self.errorhandler.logerror("Could not get translation for language %r" % language)
        return self.translation

  def refreshstats(self, args):
    """refreshes all the available statistics..."""
    if args:
      def filtererrorhandler(functionname, str1, str2, e):
        print "error in filter %s: %r, %r, %s" % (functionname, str1, str2, e)
        return False
      checkerclasses = [projects.checks.StandardChecker, projects.pofilter.StandardPOChecker]
      stdchecker = projects.pofilter.POTeeChecker(checkerclasses=checkerclasses, errorhandler=filtererrorhandler)
      for arg in args:
        if not os.path.exists(arg):
          print "file not found:", arg
        if os.path.isdir(arg):
          if not arg.endswith(os.sep):
            arg += os.sep
          projectcode, languagecode = self.potree.getcodesfordir(arg)
          dummyproject = projects.DummyStatsProject(arg, stdchecker, projectcode, languagecode)
          def refreshdir(dummy, dirname, fnames):
            reldirname = dirname.replace(dummyproject.podir, "")
            for fname in fnames:
              fpath = os.path.join(reldirname, fname)
              if fname.endswith(".po") and not os.path.isdir(os.path.join(dummyproject.podir, fpath)):
                print "refreshing stats for", fpath
                projects.pootlefile.pootlefile(dummyproject, fpath).updatequickstats()
          os.path.walk(arg, refreshdir, None)
          if projectcode and languagecode:
            dummyproject.savequickstats()
        elif os.path.isfile(arg):
          dummyproject = projects.DummyStatsProject(".", stdchecker)
          print "refreshing stats for", arg
          projects.pootlefile.pootlefile(dummyproject, arg)
    else:
      print "refreshing stats for all files in all projects"
      self.potree.refreshstats()

  def generateactivationcode(self):
    """generates a unique activation code"""
    return "".join(["%02x" % int(random.random()*0x100) for i in range(16)])

  def getpage(self, pathwords, session, argdict):
    """return a page that will be sent to the user"""
    # TODO: strip off the initial path properly
    while pathwords and pathwords[0] == "pootle":
      pathwords = pathwords[1:]
    if pathwords:
      top = pathwords[0]
    else:
      top = ""
    if top == 'js':
      pathwords = pathwords[1:]
      jsfile = os.path.join(filelocations.htmldir, 'js', *pathwords)
      if not os.path.exists(jsfile):
        jsfile = os.path.join(filelocations.jtoolkitdir, 'js', *pathwords)
        if not os.path.exists(jsfile):
          return None
      jspage = widgets.PlainContents(None)
      jspage.content_type = "application/x-javascript"
      jspage.sendfile_path = jsfile
      jspage.allowcaching = True
      return jspage
    elif pathwords and pathwords[-1].endswith(".css"):
      cssfile = os.path.join(filelocations.htmldir, *pathwords)
      if not os.path.exists(cssfile):
        cssfile = os.path.join(filelocations.jtoolkitdir, *pathwords)
        if not os.path.exists(cssfile):
          return None
      csspage = widgets.PlainContents(None)
      csspage.content_type = "text/css"
      csspage.sendfile_path = cssfile
      csspage.allowcaching = True
      return csspage
    elif top == 'images':
      pathwords = pathwords[1:]
      picturefile = os.path.join(filelocations.htmldir, 'images', *pathwords)
      picture = widgets.SendFile(picturefile)
      picture.content_type = thumbgallery.getcontenttype(pathwords[-1])
      picture.allowcaching = True
      return picture
    elif top == "testtemplates.html":
      return templateserver.TemplateServer.getpage(self, pathwords, session, argdict)
    elif not top or top == "index.html":
      return indexpage.PootleIndex(self.potree, session)
    elif top == 'about.html':
      return indexpage.AboutPage(session)
    elif top == "login.html":
      if session.isopen:
        returnurl = argdict.get('returnurl', None) or getattr(self.instance, 'homepage', 'home/')
        return server.Redirect(returnurl)
      if 'username' in argdict:
        session.username = argdict["username"]
      return users.LoginPage(session, languagenames=self.languagenames)
    elif top == "register.html":
      return self.registerpage(session, argdict)
    elif top == "activate.html":
      return self.activatepage(session, argdict)
    elif top == "projects":
      pathwords = pathwords[1:]
      if pathwords:
        top = pathwords[0]
      else:
        top = ""
      if not top or top == "index.html":
        return indexpage.ProjectsIndex(self.potree, session)
      else:
        projectcode = top
        if not self.potree.hasproject(None, projectcode):
          return None
        pathwords = pathwords[1:]
        if pathwords:
          top = pathwords[0]
        else:
          top = ""
        if not top or top == "index.html":
          return indexpage.ProjectLanguageIndex(self.potree, projectcode, session)
        elif top == "admin.html":
          return adminpages.ProjectAdminPage(self.potree, projectcode, session, argdict)
    elif top == "languages":
      pathwords = pathwords[1:]
      if pathwords:
        top = pathwords[0]
      else:
        top = ""
      if not top or top == "index.html":
        return indexpage.LanguagesIndex(self.potree, session)
    elif top == "home":
      pathwords = pathwords[1:]
      if pathwords:
        top = pathwords[0]
      else:
        top = ""
      if not session.isopen:
        templatename = "redirect"
        templatevars = {
            "pagetitle": session.localize("Redirecting to login..."),
            "refresh": 1,
            "refreshurl": "login.html",
            "message": session.localize("Need to log in to access home page"),
            }
        pagelayout.completetemplatevars(templatevars, session)
        return server.Redirect("../login.html", withtemplate=(templatename, templatevars))
      if not top or top == "index.html":
        return indexpage.UserIndex(self.potree, session)
      elif top == "options.html":
        if "changeoptions" in argdict:
          session.setoptions(argdict)
        if "changepersonal" in argdict:
          session.setpersonaloptions(argdict)
        if "changeinterface" in argdict:
          session.setinterfaceoptions(argdict)
        return users.UserOptions(self.potree, session)
    elif top == "admin":
      pathwords = pathwords[1:]
      if pathwords:
        top = pathwords[0]
      else:
        top = ""
      if not session.isopen:
        templatename = "redirect"
        templatevars = {
            "pagetitle": session.localize("Redirecting to login..."),
            "refresh": 1,
            "refreshurl": "login.html",
            "message": session.localize("Need to log in to access admin page"),
            }
        pagelayout.completetemplatevars(templatevars, session)
        return server.Redirect("../login.html", withtemplate=(templatename, templatevars))
      if not session.issiteadmin():
        templatename = "redirect"
        templatevars = {
            "pagetitle": session.localize("Redirecting to home..."),
            "refresh": 1,
            "refreshurl": "login.html",
            "message": self.localize("You do not have the rights to administer pootle."),
            }
        pagelayout.completetemplatevars(templatevars, session)
        return server.Redirect("../index.html", withtemplate=(templatename, templatevars))
      if not top or top == "index.html":
        if "changegeneral" in argdict:
          self.changeoptions(argdict)
        return adminpages.AdminPage(self.potree, session, self.instance)
      elif top == "users.html":
        if "changeusers" in argdict:
          self.changeusers(session, argdict)
        return adminpages.UsersAdminPage(self, session.loginchecker.users, session, self.instance)
      elif top == "languages.html":
        if "changelanguages" in argdict:
          self.potree.changelanguages(argdict)
        return adminpages.LanguagesAdminPage(self.potree, session, self.instance)
      elif top == "projects.html":
        if "changeprojects" in argdict:
          self.potree.changeprojects(argdict)
        return adminpages.ProjectsAdminPage(self.potree, session, self.instance)
    elif top == "templates" or self.potree.haslanguage(top):
      languagecode = top
      pathwords = pathwords[1:]
      if pathwords:
        top = pathwords[0]
        bottom = pathwords[-1]
      else:
        top = ""
        bottom = ""
      if not top or top == "index.html":
        return indexpage.LanguageIndex(self.potree, languagecode, session)
      if self.potree.hasproject(languagecode, top):
        projectcode = top
        project = self.potree.getproject(languagecode, projectcode)
        pathwords = pathwords[1:]
        if pathwords:
          top = pathwords[0]
        else:
          top = ""
        if not top or top == "index.html":
          return indexpage.ProjectIndex(project, session, argdict)
        elif top == "admin.html":
          return adminpages.TranslationProjectAdminPage(self.potree, project, session, argdict)
        elif bottom == "translate.html":
          if len(pathwords) > 1:
            dirfilter = os.path.join(*pathwords[:-1])
          else:
            dirfilter = ""
          try:
            return translatepage.TranslatePage(project, session, argdict, dirfilter)
          except projects.RightsError, stoppedby:
            argdict["message"] = str(stoppedby)
            return indexpage.ProjectIndex(project, session, argdict, dirfilter)
        elif bottom == "spellcheck.html":
          # the full review page
          argdict["spellchecklang"] = languagecode
          return spellui.SpellingReview(session, argdict, js_url="/js/spellui.js")
        elif bottom == "spellingstandby.html":
          # a simple 'loading' page
          return spellui.SpellingStandby()
        elif bottom.endswith("." + project.fileext):
          pofilename = os.path.join(*pathwords)
          if argdict.get("translate", 0):
            try:
              return translatepage.TranslatePage(project, session, argdict, dirfilter=pofilename)
            except projects.RightsError, stoppedby:
              argdict["message"] = str(stoppedby)
              return indexpage.ProjectIndex(project, session, argdict, dirfilter=pofilename)
          elif argdict.get("index", 0):
            return indexpage.ProjectIndex(project, session, argdict, dirfilter=pofilename)
          else:
            pofile = project.getpofile(pofilename, freshen=False)
            page = widgets.SendFile(pofile.filename)
            page.etag = str(pofile.pomtime)
            page.allowcaching = True
            encoding = pofile.encoding or "UTF-8"
            page.content_type = "text/plain; charset=%s" % encoding
            return page
        elif bottom.endswith(".csv") or bottom.endswith(".xlf") or bottom.endswith(".ts") or bottom.endswith("mo"):
          destfilename = os.path.join(*pathwords)
          basename, extension = os.path.splitext(destfilename)
          pofilename = basename + os.extsep + project.fileext
          extension = extension[1:]
          if extension == "mo":
            if not "pocompile" in project.getrights(session):
              return None
          etag, filepath_or_contents = project.convert(pofilename, extension)
          if etag:
            page = widgets.SendFile(filepath_or_contents)
            page.etag = str(etag)
          else:
            page = widgets.PlainContents(filepath_or_contents)
          page.allowcaching = True
          if extension == "csv":
            page.content_type = "text/plain; charset=UTF-8"
          elif extension == "xlf" or extension == "ts":
            page.content_type = "text/xml; charset=UTF-8"
          elif extension == "mo":
            page.content_type = "application/octet-stream"
          return page
        elif bottom.endswith(".zip"):
          if not "archive" in project.getrights(session):
            return None
          if len(pathwords) > 1:
            dirfilter = os.path.join(*pathwords[:-1])
          else:
            dirfilter = None
          goal = argdict.get("goal", None)
          if goal:
            goalfiles = project.getgoalfiles(goal)
            pofilenames = []
            for goalfile in goalfiles:
              pofilenames.extend(project.browsefiles(goalfile))
          else:
            pofilenames = project.browsefiles(dirfilter)
          archivecontents = project.getarchive(pofilenames)
          page = widgets.PlainContents(archivecontents)
          page.content_type = "application/zip"
          return page
        elif bottom.endswith(".sdf") or bottom.endswith(".sgi"):
          if not "pocompile" in project.getrights(session):
            return None
          oocontents = project.getoo()
          page = widgets.PlainContents(oocontents)
          page.content_type = "text/tab-seperated-values"
          return page
        elif bottom == "index.html":
          if len(pathwords) > 1:
            dirfilter = os.path.join(*pathwords[:-1])
          else:
            dirfilter = None
          return indexpage.ProjectIndex(project, session, argdict, dirfilter)
        else:
          return indexpage.ProjectIndex(project, session, argdict, os.path.join(*pathwords))
    return None

class PootleOptionParser(simplewebserver.WebOptionParser):
  def __init__(self):
    versionstring = "%%prog %s\njToolkit %s\nTranslate Toolkit %s" % (pootleversion.ver, jtoolkitversion.ver, toolkitversion.ver)
    simplewebserver.WebOptionParser.__init__(self, version=versionstring)
    self.set_default('prefsfile', filelocations.prefsfile)
    self.set_default('instance', 'Pootle')
    self.set_default('htmldir', filelocations.htmldir)
    self.add_option('', "--refreshstats", dest="action", action="store_const", const="refreshstats",
        default="runwebserver", help="refresh the stats files instead of running the webserver")

def checkversions():
  """Checks that version dependencies are met"""
  if not hasattr(toolkitversion, "build") or toolkitversion.build < 9000:
    raise RuntimeError("requires Translate Toolkit version >= 0.9.  Current installed version is: %s" % toolkitversion.ver)

def main():
  # run the web server
  checkversions()
  parser = PootleOptionParser()
  options, args = parser.parse_args()
  if options.action != "runwebserver":
    options.servertype = "dummy"
  server = parser.getserver(options)
  if options.action == "runwebserver":
    simplewebserver.run(server, options)
  elif options.action == "refreshstats":
    server.refreshstats(args)

if __name__ == '__main__':
  main()

